rclcpp/test/rclcpp/executors/test_executors.cpp

```diff
@@ -0,0 +1,448 @@
+using StandardExecutors =
+::testing::Types<
+rclcpp::executors::SingleThreadedExecutor,
+rclcpp::executors::MultiThreadedExecutor>;
+TYPED_TEST_CASE(TestExecutorsSpinVariants, StandardExecutors);
+
+// Make sure that executors detach from nodes when destructing
+TYPED_TEST(TestExecutors, detachOnDestruction) {
+  using ExecutorType = TypeParam;
+  {
+    ExecutorType executor;
+    executor.add_node(this->node);
+  }
+  {
+    ExecutorType executor;
+    EXPECT_NO_THROW(executor.add_node(this->node));
+  }
+}
+
+// Make sure that the executor can automatically remove expired nodes correctly
+TYPED_TEST(TestExecutorsSpinVariants, addTemporaryNode) {
+  using ExecutorType = TypeParam;
+  ExecutorType executor;
+  executor.add_node(std::make_shared<rclcpp::Node>("temporary_node"));
+
+  // Sleep for a short time to verify executor.spin() is going, and didn't throw.
+  std::thread spinner([&](){EXPECT_NO_THROW(executor.spin());});
+
+  std::this_thread::sleep_for(50ms);
+  rclcpp::shutdown();
+  spinner.join();
+}
+
+// Make sure that the executor can automatically remove expired nodes correctly
+TYPED_TEST(TestExecutors, addNodeTwoExecutors) {
+  using ExecutorType = TypeParam;
+  ExecutorType executor1;
+  ExecutorType executor2;
+  EXPECT_NO_THROW(executor1.add_node(this->node));
+  EXPECT_THROW(executor2.add_node(this->node), std::runtime_error);
+}
+
+// Check simple spin example
+TYPED_TEST(TestExecutors, spinWithTimer) {
+  using ExecutorType = TypeParam;
+  ExecutorType executor;
+
+  bool timer_completed = false;
+  auto timer = this->node->create_wall_timer(1ms, [&](){timer_completed=true;});
+  executor.add_node(this->node);
+
+  std::thread spinner([&](){executor.spin();});
+  std::this_thread::sleep_for(50ms);
+  EXPECT_TRUE(timer_completed);
+  rclcpp::shutdown();
+  spinner.join();
+}
+
+TYPED_TEST(TestExecutors, spinWhileAlreadySpinning) {
+  using ExecutorType = TypeParam;
+  ExecutorType executor;
+  executor.add_node(this->node);
+
+  std::thread spinner([&](){executor.spin();});
+  // Sleep for a short time to verify executor.spin() is going, and didn't throw.
+  std::this_thread::sleep_for(50ms);
+
+  EXPECT_THROW(executor.spin(), std::runtime_error);
+  rclcpp::shutdown();
+  spinner.join();
+}
+
+// Check executor exits immediately if future is complete.
+TYPED_TEST(TestExecutors, testSpinUntilFutureComplete) {
+  using ExecutorType = TypeParam;
+  ExecutorType executor;
+  executor.add_node(this->node);
+  std::future<void> future;
+
+  // test success of an immediately finishing future
+  future = std::async(
+    std::launch::async,
+    []() {
+      return;
+    });
+
+  bool spin_exited = false;
+
+  std::thread spinner([&](){
+    auto ret = executor.spin_until_future_complete(future, 1s);
+    EXPECT_EQ(rclcpp::FutureReturnCode::SUCCESS, ret);
+    spin_exited = true;
+  });
+
+  // Force the executor to do a small amount of work
+  this->publisher->publish(std_msgs::msg::Empty());
+  std::this_thread::sleep_for(1ms);
```

Comment:
This seems fragile, maybe it should check in a loop allowing for time and missed connections to occur. We're not testing the pub/sub's reliability or performance here, we're just testing that the future is completed and that spin wakes up after that. Or you could use a timer instead.

Reply:
Replaced it with a for loop like I have elsewhere. 

I did have a question about the spin_until_future_complete API. Is it intentional that if the executor does not have any work to do (no active subscriptions or whatnot), then the `spin_once_impl` call inside `spin_until_future_complete` can block until the timeout (or indefinitely if no timeout is used) even if the future completed? It seems like that goes against what spin_until_future_complete's name actually implies.

Reply:
Yeah, that is intentional. It is expected that the future is completed by the callback associated with one of the entities being waited on. This isn't documented though. The only way to handle it differently would be to poll the state of the future. We cannot wait on `rmw_wait` and a condition variable simultaneously.